tests/cancun/eip1153_tstore/test_tstorage_reentrancy_contexts.py

```diff
@@ -0,0 +1,218 @@
+                "call. TSTORE(x, y), CALL(self, ...), TSTORE(x, z), TSTORE(x, z + 1) REVERT, ",
+                "TLOAD(x) returns y.",
+                "",
+                "Based on [ethereum/tests/.../09_revertUndoesAllFiller.yml](https://github.com/ethereum/tests/blob/9b00b68593f5869eb51a6659e1cc983e875e616b/src/EIPTestsFiller/StateTests/stEIP1153-transientStorage/09_revertUndoesAllFiller.yml).",  # noqa: E501
+            ),
+            "callee_bytecode": Conditional(
+                condition=SETUP_CONDITION,
+                # setup
+                if_true=(
+                    Op.TSTORE(0xFE, 0x100)
+                    + Op.TSTORE(0xFF, 0x101)
+                    + REENTRANT_CALL
+                    + Op.SSTORE(1, Op.TLOAD(0xFE))  # test value not updated during reentrant call
+                    + Op.SSTORE(2, Op.TLOAD(0xFF))  # test value not updated during reentrant call
+                ),
+                # reenter
+                if_false=(
+                    # store twice and revert; none of the stores should take effect
+                    Op.TSTORE(0xFE, 0x201)
+                    + Op.TSTORE(0xFE, 0x202)
+                    + Op.TSTORE(0xFF, 0x201)
+                    + Op.TSTORE(0xFF, 0x202)
+                    + Op.REVERT(0, 0)
+                ),
+            ),
+            "expected_callee_storage": {0: 0x00, 1: 0x100, 2: 0x101},
+        },
+    )
+
+    REVERT_UNDOES_TSTORAGE_AFTER_SUCCESSFUL_CALL = {
+        "description": (
+            "Revert undoes transient storage writes from inner calls that successfully returned. ",
+            "TSTORE(x, y), CALL(self, ...), CALL(self, ...), TSTORE(x, y + 1), RETURN, REVERT, "
+            "TLOAD(x) returns y."
+            "",
+            "Based on stEIP1153-transientStorage/10_revertUndoesStoreAfterReturnFiller.yml",
+        ),
+        "callee_bytecode": Conditional(
+            condition=SETUP_CONDITION,
+            # setup
+            if_true=(
+                Op.TSTORE(0xFF, 0x100)
+                + Op.SSTORE(2, Op.TLOAD(0xFF))
+                + Op.MSTORE(0, 2)
+                + Op.SSTORE(0, Op.CALL(Op.GAS(), callee_address, 0, 0, 32, 0, 32))
+                + Op.SSTORE(1, Op.MLOAD(0))  # should be 1 (successful call)
+                + Op.SSTORE(3, Op.TLOAD(0xFF))
+            ),
+            # first, reentrant call, which reverts
+            if_false=Conditional(
+                condition=Op.EQ(Op.CALLDATALOAD(0), 0x02),
+                if_true=(
+                    Op.MSTORE(0, 3)
+                    + Op.MSTORE(0, Op.CALL(Op.GAS(), callee_address, 0, 0, 32, 0, 0))
+                    + Op.REVERT(0, 32)
+                ),
+                # second, successful reentrant call
+                if_false=Op.TSTORE(0xFF, 0x101),
+            ),
+        ),
+        "expected_callee_storage": {0: 0x00, 1: 0x01, 2: 0x100, 3: 0x100},
+    }
+
+    def __init__(self, test_case):
```

Comment:
These are all brilliant :100: 

Might be worth it to add a copy of the `REVERT` ones but using `INVALID` opcode, to test that it has the same effect, and also using less gas when calling to not run out of gas on the calling frame.
Perhaps even adding another variant with an infinite loop to actually test an Out-of-Gas condition.

Reply:
Done in 530b5895, but the "invalid version" (`invalid_undoes_tstorage_after_successful_call`) of `revert_undoes_tstorage_after_successful_call` is currently marked to xfail. Need to change how we retrieve the return value of the CALL from the subcontext to the main context.